Objects/typeobject.c

```diff
@@ -2241,6 +2241,12 @@ extra_ivars(PyTypeObject *type, PyTypeObject *base)
         type->tp_weaklistoffset + sizeof(PyObject *) == t_size &&
         type->tp_flags & Py_TPFLAGS_HEAPTYPE)
         t_size -= sizeof(PyObject *);
+    if (!(type->tp_flags & Py_TPFLAGS_MANAGED_DICT) &&
```

Comment:
We need to check for `tp_dictoffset` and `tp_weaklistoffset`, and shouldn't presume the order.
So we need to check for weakrefs, then dict, then weakrefs again (or vice versa).
```C
if (FIELD_AT_OFFSET(dictoffset, t_size)) {
   t_size -= sizeof(PyObject *);
}
if (FIELD_AT_OFFSET(weaklistoffset, t_size)) {
   t_size -= sizeof(PyObject *);
}
if (FIELD_AT_OFFSET(dictoffset, t_size)) {
   t_size -= sizeof(PyObject *);
}
```